src/matching/Generic_vs_generic.ml

```diff
@@ -406,14 +415,21 @@ let m_with_symbolic_propagation ~is_root f b =
            *
            * nosemgrep *)
           if b1 == b then (
-            logger#warning
+            logger#error
               "Aborting symbolic propagation: Circular reference encountered \
                (\"%s\")"
               id;
-            fail ())
-          else f b1
-      | ___else___ -> fail ())
-    ~else_:(fail ())
+            fail () tin)
+          else (
+            incr tin.followed_sym_vals;
```

Comment:
I'd be more comfortable if this weren't mutable, and we instead allocated a new `tin` with the updated value. I think it would make it a lot more clear that the limit has to do with the depth of the symbolic propagation. Right now, I would have to understand exactly when we allocate a new `tin` in order to understand at what point we give up on symbolic propagation.

I'd probably have to read quite a bit of code to convince myself that this wouldn't cause a spurious failure when we perform a lot of unrelated symbolic propagation operations as part of one match attempt.

Is there a specific reason that you've done it this way? If not, I'd appreciate it if you switched it to be immutable.

Reply:
Good point! Initially I tried to make it immutable but it looked ugly, I think I just overcomplicated myself. I gave it another try after your comment and I got it done in a simple way. 🤷‍♂️